src/com/facebook/buck/android/DefaultAndroidLibraryCompilerFactory.java

```diff
@@ -32,12 +37,68 @@ public DefaultAndroidLibraryCompilerFactory(
   }
 
   @Override
-  public AndroidLibraryCompiler getCompiler(AndroidLibraryDescription.JvmLanguage language) {
+  public AndroidLibraryCompiler getCompiler(AndroidLibraryDescription.Arg args) {
+    JvmLanguage language = null;
+
+    // We currently expect all non-java languages to only contain files of that extension.
+    for (SourcePath sourcePath : args.srcs.get()) {
+      if (sourcePath instanceof PathSourcePath) {
+        String extension = Files.getFileExtension(sourcePath.toString()).toLowerCase();
+        if (extension.contains(KotlinBuckConfig.EXTENSION)) {
+          if (language == null) {
+            language = JvmLanguage.KOTLIN;
+          } else if (language != JvmLanguage.KOTLIN) {
+            throwMismatchedSourceException(language, extension);
+          }
+        } else if (extension.contains(ScalaBuckConfig.EXTENSION)) {
+          if (language == null) {
+            language = JvmLanguage.SCALA;
+          } else if (language != JvmLanguage.SCALA) {
+            throwMismatchedSourceException(language, extension);
+          }
+        }
+      }
+    }
+
+    if (language == null) {
+      language = JvmLanguage.JAVA;
+    }
+
+    switch(language) {
+      case JAVA:
+        return new JavaAndroidLibraryCompiler();
+      case KOTLIN:
+        return new KotlinAndroidLibraryCompiler(kotlinBuckConfig);
+      case SCALA:
+        return new ScalaAndroidLibraryCompiler(scalaConfig);
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  private void throwMismatchedSourceException(JvmLanguage language, String extension) {
+    String previousExtension = "unknown";
     switch (language) {
-      case JAVA: return new JavaAndroidLibraryCompiler();
-      case SCALA: return new ScalaAndroidLibraryCompiler(scalaConfig);
-      case KOTLIN: return new KotlinAndroidLibraryCompiler(kotlinBuckConfig);
+      case JAVA:
+        previousExtension = JavaBuckConfig.EXTENSION;
+        break;
+      case KOTLIN:
+        previousExtension = KotlinBuckConfig.EXTENSION;
+        break;
+      case SCALA:
+        previousExtension = ScalaBuckConfig.EXTENSION;
+        break;
     }
-    throw new HumanReadableException("Unsupported `language` parameter value: %s", language);
+    throw new HumanReadableException(
+        "Non-java jvm languages can only have sources of one filetype. Found .%s and .%s files",
```

Comment:
This seems limiting. Why can the sourcepaths not be compiled separately and combined together later? Gradle supports this already for example